diagnostics/collector.go

```diff
@@ -0,0 +1,170 @@
+    , pg_size_pretty(toast_bytes) AS toast
+    , pg_size_pretty(table_bytes) AS TABLE
+  FROM (SELECT *, total_bytes-index_bytes-COALESCE(toast_bytes,0) AS table_bytes FROM (
+      SELECT c.oid,nspname AS table_schema, relname AS TABLE_NAME
+              , c.reltuples AS row_estimate
+              , pg_total_relation_size(c.oid) AS total_bytes
+              , pg_indexes_size(c.oid) AS index_bytes
+              , pg_total_relation_size(reltoastrelid) AS toast_bytes
+          FROM pg_class c
+          LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+          WHERE relkind = 'r') a) a`)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("get table sizes %w", err))
+	}
+
+	// write all errors into log
+	if err := c.writeErrors(ctx, writer, errs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Collector) writeQueryContent(ctx context.Context, writer *zip.Writer, name string, selectBuilder sq.SelectBuilder) error {
+	q := selectBuilder.Limit(maxRecordsPerQueryCollect)
+	query, args, err := q.ToSql()
+	if err != nil {
+		return fmt.Errorf("query build: %w", err)
+	}
+	return c.writeRawQueryContent(ctx, writer, name, query, args...)
+}
+
+func (c *Collector) writeRawQueryContent(ctx context.Context, writer *zip.Writer, name string, query string, args ...interface{}) error {
+	rows, err := c.db.Query(query, args...)
+	if err != nil {
+		return fmt.Errorf("execute query: %w", err)
+	}
+	defer rows.Close()
+
+	filename := name + csvFileExt
+	w, err := writer.Create(filename)
+	if err != nil {
+		return fmt.Errorf("new file for table %s - %w", name, err)
+	}
+	csvWriter := csv.NewWriter(w)
+	defer csvWriter.Flush()
+
+	first := true
+	for rows.Next() {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		if first {
+			first = false
+			descriptions := rows.FieldDescriptions()
+			if len(descriptions) == 0 {
+				return ErrNoColumnsFound
+			}
+			cols := make([]string, len(descriptions))
+			for i, fd := range descriptions {
+				cols[i] = string(fd.Name)
+			}
+			if err := csvWriter.Write(cols); err != nil {
+				return fmt.Errorf("write csv header for %s %w", name, err)
+			}
+		}
+		v, err := rows.Values()
+		if err != nil {
+			return err
+		}
+		record := make([]string, len(v))
+		for i := range v {
+			record[i] = fmt.Sprintf("%v", v[i])
+		}
+		if err := csvWriter.Write(record); err != nil {
+			return err
+		}
+	}
+	return rows.Err()
+}
+
+func (c *Collector) writeTableContent(ctx context.Context, writer *zip.Writer, name string) error {
+	q := sq.Select("*").From(name)
+	return c.writeQueryContent(ctx, writer, name, q)
+}
+
+func (c *Collector) writeTableCount(ctx context.Context, writer *zip.Writer, name string) error {
+	q := sq.Select("COUNT(*)").From(name)
+	return c.writeQueryContent(ctx, writer, name+"_count", q)
+}
+
+func (c *Collector) writeErrors(ctx context.Context, writer *zip.Writer, errs []error) error {
+	if len(errs) == 0 {
+		return nil
+	}
+	w, err := writer.Create("errors.log")
```

Comment:
If any error occurred, you'll always be writing 2 files, the `output.log` and the `errors.log`. It would be nice if they could be written to the same folder.

Reply:
what is output.log?

Reply:
Sorry, I meant 'lakefs-diagnostics.zip', it's better if it's always next to the errors, no?

Reply:
want to get a zip file and if some of the queries didn't run - I want to see it in the errors.txt found there.

Reply:
Gotcha